hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java

```diff
@@ -22,13 +27,22 @@
 
 /**
  * @author Steve Ebersole
+ * @author Sanne Grinovero
  */
 class EventListenerGroupImpl<T> implements EventListenerGroup<T> {
 	private EventType<T> eventType;
 	private final EventListenerRegistryImpl listenerRegistry;
 
 	private final Set<DuplicationStrategy> duplicationStrategies = new LinkedHashSet<>();
-	private List<T> listeners;
+
+	// Performance: make sure iteration on this type is efficient; in particular we do not want to allocate iterators,
+	// not having to capture state in lambdas.
+	// So we keep the listeners in both a List (for convenience) and in an array (for iteration). Make sure
+	// their content stays in synch!
+	private T[] listeners = null;
```

Comment:
we can save one JVM step by avoiding assignment of default value. JVM has incurred overhead by assigning any reference field to null just after object is allocated.